### ISSUE TITLE
DELETE requests to deletef and forwardf routing to a controller fail with 404 not found

### ISSUE THREAD

**Issue by MarneeDear** [SaturnFramework/Saturn#160]
Possibly related to issue #54 

Using the latest from Saturn master branch as of 12/20/2018 at 10:00 MDT.

Using `Controller.Sample` I added a `deletef` route to `topRouter` and added a new `deleteRoute` router.  The route pattern is `/delete/%i`. 

This is the URL I put in PostMan to make the DELETE request: `http://localhost:8085/delete/1111`

You can see the source code for this in my forked repo here: https://github.com/MarneeDear/Saturn/blob/delete-routes-404/sample/Controller.Sample/Controller.Sample.fs#L105

## Delete requests to this route fail with 404 when the route sends directly to a controller
```fsharp
   deletef "/delete/%i" (fun (_ : int) -> userController)
```
## Requests directly to a function are OK.

```fsharp
let apiDeleteExample2 id = sprintf "Echo: %i" id |> text

//inside topRouter
    deletef "/delete/%s" apiDeleteExample2
```
I'm using dotnet on Windows 10.
When I run Controller.Sample I do `dotnet watch run`.

## UPDATE
I think this happens when the `id` passed to the controller is not a `string`. 

### This doesnt work (`id` is type `int`)
```fsharp

//In userController
    delete (fun ctx id -> (sprintf "Delete handler no version failed - %i" id) |> Controller.text ctx)

//In topRouter
    deletef "/delete/%i" (fun (_:int) -> userController)
```

### This works (`id` is type `string`)

```fsharp
//In userController
    delete (fun ctx id -> (sprintf "Delete handler no version failed - %s" id) |> Controller.text ctx)

//In topRouter
   deletef "/delete/%s" (fun (_:string) -> userController)
```

**Comment by MarneeDear** [comment]
`forwardf` instead of `getf` seems to work, but I am having problems with pointing it to another router. I'm going to write a new issue.

**Comment by MarneeDear** [comment]
Made the comment on the wrong issue.

**Comment by MarneeDear** [comment]
I suspect I will need to use forwardf. I am trying that out.

**Comment by MarneeDear** [comment]
DELETE to a `forwardf` to a controller are 404.

These both 404.

```
    deletef "/delete/%i" (fun (_:int) -> userController)
    forwardf "/delete/%i" (fun (_ : int) -> userController)
```
 See full source code here.
https://github.com/MarneeDear/Saturn/blob/delete-routes-404/sample/Controller.Sample/Controller.Sample.fs#L112

**Comment by NinoFloris** [comment]
I just looked at what the code is doing, and its not wrong. However it's not obvious how it all works either. To understand you need to know a few important points to grasp why all your examples do what they do.

- Saturn uses Giraffe subrouting under the hood, only for `forward` and `forwardf` calls as mention with jargon in their method details.
```Forwards calls to different scope. Modifies the HttpRequest.Path to allow subrouting.```

     When you use `forward/forwardf`, they consume their matched route segment so to say, allowing the next handler to start matching at the subsequent segment. This is how `subController` works as well.

- If you use any consuming 'formatting' (so `forwardf` and `subController`) route handler, always use the passed-in argument. This is the only time you'll get that route argument. So every example that does something like: 
```deletef "/delete/%i" (fun (_:int) -> userController)```
will discard the %i and put the route pointer past that segment. So by the time the userController matches its action routes they all start their matches AFTER `/delete/%i`. 

    So that means in this case when you would call it with `delete/11/1` you would get `1` as the id in your controller actions. For `delete/11` it correctly 404s if you don't have an index/post/deleteall action.

Now you also see, when you don't use a consuming route handler as delete/deletef and related friends are, every subsequent handler will start their match at the same point in path again. 

In your samples you write out `id` as `%s` in your `sprintf` handler. Through F#s type inference that will set the controller's `'Key` type as string. The result is that the string route basically matches anything of path, returning to you `delete/11` as the id. 

This last part feels like a bug though, and is something I need to think about, it seems incorrect that a `'Key = string` controller would just always match the full remaining path, intuitively I would expect it would return `delete` as the id instead, stopping at the first unencoded '/'. What do you think @Krzysztof-Cieslak?

Hope this helps!

**Comment by NinoFloris** [comment]
Fix for weird matching behavior of %s controllers is pushed, see https://github.com/SaturnFramework/Saturn/pull/168

**Comment by Krzysztof-Cieslak** [comment]
Hey, @NinoFloris thanks for the detailed response!

@MarneeDear, I'll close this issue - as Nino mentioned above, most of the behaviour is expected, and the `%s` bug was fixed by #168 and will be available in next Saturn release. Please let us know if we can help more with the problems - we probably need to clarify some documentation around (sub)routing.
